openstack/compute/v2/extensions/remoteconsoles/requests.go

```diff
@@ -0,0 +1,83 @@
+package remoteconsoles
+
+import (
+	"github.com/gophercloud/gophercloud"
+)
+
+// RemoteConsoleProtocol represents valid remote console protocol.
+// It can be used to create a remote console with one of the pre-defined protocol.
+type RemoteConsoleProtocol string
+
+const (
+	// RemoteConsoleVNCProtocol represents the VNC console protocol.
+	RemoteConsoleVNCProtocol RemoteConsoleProtocol = "vnc"
+
+	// RemoteConsoleSPICEProtocol represents the SPICE console protocol.
+	RemoteConsoleSPICEProtocol RemoteConsoleProtocol = "spice"
+
+	// RemoteConsoleRDPProtocol represents the RDP console protocol.
+	RemoteConsoleRDPProtocol RemoteConsoleProtocol = "rdp"
+
+	// RemoteConsoleSerialProtocol represents the serial console protocol.
+	RemoteConsoleSerialProtocol RemoteConsoleProtocol = "serial"
+
+	// RemoteConsoleMKSProtocol represents the MKS console protocol.
+	RemoteConsoleMKSProtocol RemoteConsoleProtocol = "mks"
+)
+
+// RemoteConsoleType represents valid remote console type.
+// It can be used to create a remote console with one of the pre-defined type.
+type RemoteConsoleType string
+
+const (
+	// RemoteConsoleNoVNCType represents the VNC console type.
+	RemoteConsoleNoVNCType RemoteConsoleType = "novnc"
+
+	// RemoteConsoleXVPVNCType represents the XVP VNC console type.
+	RemoteConsoleXVPVNCType RemoteConsoleType = "xvpvnc"
+
+	// RemoteConsoleRDPHTML5Type represents the RDP HTML5 console type.
+	RemoteConsoleRDPHTML5Type RemoteConsoleType = "rdp-html5"
+
+	// RemoteConsoleSPICEHTML5Type represents the SPICE HTML5 console type.
+	RemoteConsoleSPICEHTML5Type RemoteConsoleType = "spice-html5"
+
+	// RemoteConsoleSerialType represents the serial console type.
+	RemoteConsoleSerialType RemoteConsoleType = "serial"
+
+	// RemoteConsoleWebMKSType represents the web MKS console type.
+	RemoteConsoleWebMKSType RemoteConsoleType = "webmks"
+)
+
+// CreateOptsBuilder allows to add additional parameters to the Create request.
+type CreateOptsBuilder interface {
+	ToRemoteConsoleCreateMap() (map[string]interface{}, error)
+}
+
+// CreateOpts specifies parameters to the Create request.
+type CreateOpts struct {
+	// Protocol specifies the protocol of a new remote console.
+	Protocol string `json:"protocol" required:"true"`
```

Comment:
Should this be:

```go
Protocol RemoteConsoleProtocol `json:"protocol" reqauired:"true"`
```

Reply:
It looks like the same issue as we discussed in [PR#1147](https://github.com/gophercloud/gophercloud/pull/1147#issuecomment-407868998).

Do you think that we need to strict users to pre-defined values in case of this change? I don't actually get how it's different from the #1147 😅

Reply:
No worries - since it's not actually written anywhere, it's hard to grasp :)

The rule is: custom types for "input", standard types for "output". So any type of `CreatOpts`, `UpdateOpts`, etc should have types which can provide better/safer input. Result structures should only contain "standard" (`string`, `int`, `[]string`, etc fields). If a struct contains a struct, then that can be a defined struct.

Reply:
Alright, seems legit.